lib/base_service.ts

```diff
@@ -245,4 +321,49 @@ export class BaseService {
     return _credentials;
   }
 
+  /**
+   * Determines if tokens are being used for authentication or not
+   * @private
+   * @returns {boolean}
+   */
+  private notUsingTokens(): boolean {
+    // if tokens are not supported, tokenInfo will be undefined
+    // if tokens are supported but the user chose to pass in
+    // username and password, use basic auth
+    return !supportsIamTokens(this.tokenInfo) || !this._options.platform_api_key;
+  }
+
+  /**
+   * Determines if an access token is already stored
+   * @private
+   * @returns {boolean}
+   */
+  private needToRequestToken(): boolean {
+    return !this.tokenInfo.access_token;
+  }
+
+  /**
+   * Saves the token response data to the object state
+   * @private
+   * @param {TokenData} tokenResponse
+   * @returns {void}
+   */
+  private setTokenObject(tokenResponse: TokenData): void {
+    this.tokenInfo = extend({}, tokenResponse);
+  }
+
+  /**
+   * Checks if token is nearing its expiration time
+   * The token should be refreshed if past a certain percentage of its time to live
+   * @private
+   * @returns {boolean}
+   */
+  private tokenIsExpired(): boolean {
+     // use a buffer to prevent the edge case of the token
+     // expiring before the request could be made
+     // the buffer will be a fraction of the total TTL
+    const fractionOfTtl = 0.8;
+    return this.tokenInfo.expiration * fractionOfTtl
+      < Math.floor(Date.now() / 1000);
```

Comment:
Shouldn't this be 80% of `expires_in` instead of 80% of `expiration`? I think 80% of `expiration` would be 80% of the time since the epoch.

I think you want to calculate: `Date.now() > request_time + 0.8 * expires_in`

But since you may not have the time when the token request was made, you could instead calculate:
`Date.now() > expiration - 0.2 * expires_in`

Reply:
Great catch, I must not have been thinking too hard when I switched it from an absolute subtraction to a multiplier.